File: structures/c/linked-list.c
#include <stdlib.h>
#include <stdio.h>

#include "linked-list.h"

List* listInitialize(ListNode* node) {
	List* list = malloc(sizeof(List));

	if (!node) {
		list -> head = NULL;
		list -> tail = NULL;
		list -> size = 0;
	} else {
		list -> head = node;
		list -> tail = node;
		list -> size = 1;
	}

	return list;
}

void listDeconstruct(List* list) {
	ListNode* current = list -> head;

	while (current != list -> tail) {
		ListNode* next = current -> next;

		free(current);

		current = next;
	}

	free(list);
	list = NULL;
}

ListNode* listNodeInitialize(int data, ListNode* next) {
	ListNode* node = malloc(sizeof(ListNode));

	node -> data = data;
	node -> next = next ? next : NULL;

	return node;
}

void listNodeDeconstruct(ListNode* node) {
	free(node);
}

void listPrint(List* list) {
	ListNode* current = list -> head;

	while (current != NULL) {
		printf("%d -> ", current -> data);
		current = current -> next;
	}

	printf("\n");
}

void listPush(List* list, int data) {
	ListNode* tail = list -> tail;

	ListNode* newElem = listNodeInitialize(data, NULL);

	tail -> next = newElem;
	list -> tail = newElem;
	list -> size++;
}

ListNode* listPop(List* list) {
	ListNode* tail = list -> tail;

	ListNode* current = list -> head;
	while (current -> next -> next != NULL) {
		current = current -> next;
	}

	current -> next = NULL;
	list -> tail = current;
	list -> size--;

	return tail;
}

ListNode* listGetElement(List* list, int index) {
	if (index >= list -> size) {
		printf("Element not found at index %d - outside of range\n", index);
		return NULL;
	}

	ListNode* node = list -> head;

	int i = 0;
	while (i < index) {
		if (node != NULL) {
			i++;
			node = node -> next;
		} else {
			printf("Element not found at index %d outside of range\n", index);
			return NULL;
		}
	}

	return node;
}

void listSetElement(List* list, int index, int newData) {
	ListNode* atIndex = listGetElement(list, index);

	if (atIndex == NULL) {
		printf("Not setting - aborting...\n");
		return;
	}

	atIndex -> data = newData;
}

void listInsert(List* list, int index, int newData) {
	ListNode* atIndex = listGetElement(list, index - 1);

	ListNode* newNode = listNodeInitialize(newData, atIndex -> next);
	atIndex -> next = newNode;

	list -> size++;
}

void listDelete(List* list, int index) {
	ListNode* prevIndex = listGetElement(list, index - 1);
	ListNode* temp = prevIndex -> next;

	prevIndex -> next = prevIndex -> next -> next;

	free(temp);
	list -> size--;
}

void listClear(List* list) {
	ListNode* current = list -> head;
	ListNode* tempHolder[list -> size];

	int i = 0;
	while (current != NULL) {
		tempHolder[i] = current;
		current = current -> next;
		i++;
	}

	for (int i = 0; i < list -> size; i++) {
		free(tempHolder[i]);
	}

	list -> head = NULL;
	list -> tail = NULL;
	list -> size = 0;
}

File: structures/python/linked_list.py
"""
Singly linked list
"""


class LinkedList:
    def __init__(self, head=None):
        self.head = head
        self.tail = head
        self.size = 1 if head else 0

    def _check_range(self, index):
        if index < 0 or index > max(self.size - 1, 0):
            raise Exception('Index out of range')

    def append(self, data):
        node = LinkedListNode(data=data)

        if not self.size:
            self.head = node
            self.tail = self.head
        else:
            self.tail.next = node
            self.tail = self.tail.next

        self.size += 1

    def pop(self, node=False):
        if self.size == 0:
            return

        if self.size == 1:
            current = self.head
            self.head = None
            self.tail = None
        else:
            previous = self.get(self.size - 2, node=True)
            current = previous.next
            self.tail = previous
            previous.next = None

        self.size -= 1
        return current if node else current.data

    def get(self, index, node=False):
        self._check_range(index)

        current = self.head

        while index > 0 and current.next is not None:
            index -= 1
            current = current.next

        if index != 0:
            raise Exception('Not Found')

        return current if node else current.data

    def find(self, data, fn=None):
        current = self.head

        for index in range(self.size):
            if not fn:
                if current.data == data:
                    return (index, current)
            else:
                if fn(current.data, data):
                    return (index, current)

            current = current.next

        return (-1, None)

    def insert(self, index, data):
        # head
        if index == 0:
            next = self.head
            self.head = LinkedListNode(data=data)
            self.tail = self.head
        #tail
        elif index == self.size:
            size = self.size
            self.append(data)
            self.size = size
        # other
        else:
            self._check_range(index)

            previous = self.get(index - 1, node=True)
            current = LinkedListNode(data=data)
            current.next = previous.next
            previous.next = current

        self.size += 1

    def delete(self, index):
        # head
        if index == 0:
            current = self.head
            self.head = self.head.next
            # remove from scope
            del current
        #tail
        elif index == self.size - 1:
            size = self.size
            self.pop(index)
            self.size = size
        # other
        else:
            self._check_range(index)

            previous = self.get(index, node=True)
            current = previous.next
            previous.next = current.next
            # remove from scope
            del current

        self.size -= 1

    def print(self, fn=None):
        current = self.head
        while current is not None:
            if fn:
                print(fn(current.data), end=' -> ')
            else:
                print(current.data, end=' -> ')
            current = current.next
        print()


class LinkedListNode:
    def __init__(self, data=None, next=None):
        self.next = next
        self.data = data
